[FLINK-6013][metrics] Add Datadog HTTP metrics reporter

bowenli86 opened the pull request:
I'm adding a DatadogHttpReporter for Flink metrics system.

The implementation, including making parameters in Flink's metrics as Datadog tags, is a best practice based on our long time working experience and understanding of Datadog. It might be a bit different than how other metrics reporters work, but it truly helps developers to find and filter metrics quickly, better categorize metrics, and visualize them on Datadog dashboards, especially when users (like we OfferUp) have a dozen individual Flink clusters.

------

Thanks for contributing to Apache Flink. Before you open your pull request, please take the following check list into consideration.
If your changes take all of the items into account, feel free to open your pull request. For more information and/or questions please refer to the [How To Contribute guide](http://flink.apache.org/how-to-contribute.html).
In addition to going through the list, please provide a meaningful description of your changes.

- [x] General
  - The pull request references the related JIRA issue ("[FLINK-6013] Add Datadog HTTP metrics reporter")
  - The pull request addresses only one issue
  - Each commit in the PR has a meaningful commit message (including the JIRA id)

- [x] Documentation
  - Documentation has been added for new functionality
  - Old documentation affected by the pull request has been updated
  - JavaDoc for public methods has been added

- [x] Tests & Build
  - Functionality added by the pull request is covered by tests
  - `mvn clean verify` has been executed successfully locally or a Travis build has passed

bowenli86 commented:
Great advice. Thanks, @zentol !  

My questions to your suggestions:
1. If I make DatadogHttpReport implement MetricReporter interface, shall I just ignore Histogram and Meter which are not supported by Datadog Http API and document that?
2. For creating DGauges/DCounters. I believe you are correct. I can create new objects upon metrics registration, and continue update their values, right?
3. About parsing tags. Great to know I can do it via MetricGroup! Thus, shall I just document the metrics scope configs should be as the following? 

```
metrics.scope.jm: jobmanager
metrics.scope.jm.job: jobmanager.job
metrics.scope.tm: taskmanager
metrics.scope.tm.job: taskmanager.job
metrics.scope.task: task
metrics.scope.operator: operator
```

zentol commented on docs/monitoring/metrics.md:
```diff
@@ -436,6 +436,38 @@ metrics.reporter.stsd.port: 8125
+
+{% endhighlight %}
+
+Such metric reporting implementation is a best practice based on our experience working with Datadog. It helps 
```
Please remove this, it doesn't belong in the documentation.

bowenli86 replied:
removed

zentol commented:
If you want to enforce a specific scope config then i would suggest to instead ignore the configured one (which means not using `MetricGroup#getMetricIdentifier()`) and instead extracting the variables you want from `MetricGroup#getAllVariables()`. The JMXReporter does this as well and is one example of a reporter that completely ignores scope formats. It builds a series key-value pairs containing the variables for every metric, not unlike this reporter.

You should be able to support Meters, as the DCounter works with Numbers. Only exporting the rate is perfectly reasonable. You don't necessarily have to ignore Histograms; you can instead create a number of DCounters for every property that is extracted from the HistogramStatistics; it's a bit more work but doable imo.

Instead of storing the actual value in the DMetric why not store the Flink metric instead? When `DMetric#getPoints()` is called you could retrieve the value form the metric, you wouldn't have to update anything manually.

zentol commented on flink-metrics/flink-metrics-datadog/pom.xml:
```diff
@@ -0,0 +1,85 @@
+		</dependency>
+
+		<dependency>
+			<groupId>com.google.guava</groupId>
```
You should shade this away to make sure there aren't any dependency conflicts.

bowenli86 replied:
shaded

zentol replied:
setting it to provided isn't shading. Shading refers to the relocation of a dependency to a different namespace using the `maven-shade-plugin`, which allows different versions of guava to be present on the class path. Basically, rewrite you rewrite all package and import declarations and include the modified version in the jar.

Do okhttp or okio rely on guava as well? If not i would propose removing the version tag from the dependency, so that Flink doesn't depend on different versions.

bowenli86 replied:
I removed guava dependency as Stephan proposed

zentol commented on flink-metrics/flink-metrics-datadog/src/main/java/org/apache/flink/metrics/datadog/metric/DGauge.java:
```diff
@@ -0,0 +1,30 @@
+/**
+ * Mapping of gauge between Flink and Datadog
+ * */
+public class DGauge<T extends Number> extends DMetric {
```
What's the benefit in differentiating between DCounters and DGauges?

bowenli86 replied:
Their MetricType are different. One is MetricType.counter and another is MetricType.gauge, which will be serialized as a json field

zentol replied:
Are counters and gauges specific types that DataDog supports? GI'm asking sinec the MetricType is a custom enum, and in no connected to any form of DataDog API.

bowenli86 replied:
yes, see my explanation below

bowenli86 commented:
Addressed @zentol 's comments. Ready for another round. Thanks!

zentol commented on flink-metrics/flink-metrics-datadog/src/main/java/org/apache/flink/metrics/datadog/DMetric.java:
```diff
@@ -0,0 +1,65 @@
+ * Abstract metric of Datadog for serialization
+ * */
+public abstract class DMetric {
+	private final String metric; // Metric name
```
I would rename this to "name", and `getMetric()` to `getName()`.

bowenli86 replied:
'metric' is a field name in serialized json, and Jackson takes the name here during serialization. Thus, I can't rename it to something else.

zentol replied:
So this is a requirement by datadog then? Which expects the name field to be called "metric"?

bowenli86 replied:
yes, exactly

zentol commented on flink-metrics/flink-metrics-datadog/src/main/java/org/apache/flink/metrics/datadog/DatadogHttpReporter.java:
```diff
@@ -0,0 +1,216 @@
+
+
+/**
+ * Abbr: dghttp
```
What's this?

bowenli86 replied:
This is reminder for flink-conf.yaml when users choose to use this reporter.
```
metrics.reporters: dghttp
```

I'll add more comment here

zentol commented on flink-metrics/flink-metrics-datadog/src/main/java/org/apache/flink/metrics/datadog/metric/MetricType.java:
```diff
@@ -0,0 +1,26 @@
+ * Metric types supported by Datadog
+ * */
+public enum MetricType {
+	gauge, counter
```
by convention enum values should be upper-case.

bowenli86 replied:
yes, 'counter' and 'gauge' are specific types in Datadog, and these will be fields `"type":"gauge"` and `"type":"counter"` in json sent to Datadog. 

Since they are meant for serialization mentioned above, I don't see any benefits make them upper-case, or doing something like `GAUGE("gauge"), COUNTER("counter")`

StephanEwen commented:
A few more comments on this:

  - Guava is so conflict heavy that should avoid using it in the framework wherever possible. Adding a multiple MBs dependency to write `Lists.newArrayList` rather than `new ArrayList<>()` also seems a bit much. The use of the `Joiner` can also be replaced with a `StringBuilder`.

  - You can drop the `logback-test.xml` - I think this is an uneeded artifact that some other modules still have for historic reasons.

  - Would be good to also make sure this is put into the `opt` folder in the build target, so it can be easily dropped into the `lib` folder.

bowenli86 commented:
Addressed @zentol and @StephanEwen 's comments. Ready for another round!

zentol commented on flink-metrics/flink-metrics-datadog/src/main/java/org/apache/flink/metrics/datadog/DatadogHttpReporter.java:
```diff
@@ -0,0 +1,214 @@
+
+	@Override
+	public void notifyOfAddedMetric(Metric metric, String metricName, MetricGroup group) {
+		final String name = group.getMetricIdentifier(metricName, this);
```
Since you don't filter characters anyway you can use `MetricGroup#getMetricIdentifier(String name)`. Then you also don't need to imp0lement `CharacterFilter`.

bowenli86 replied:
make sense. Refactored as you suggested.

btw, what's CharacterFilter#filterCharacters() for?

zentol replied:
Some systems don't allow certain special characters to be part of the metric name. Instead of filtering these out after generating the MetricIdentifier every time, the CharacterFilter interface allows us to do this filtering more efficiently by caching a filtered version of the scope in the MetricGroup.

zentol commented on flink-metrics/flink-metrics-datadog/src/main/java/org/apache/flink/metrics/datadog/DatadogHttpReporter.java:
```diff
@@ -0,0 +1,214 @@
+		List<String> tags = new ArrayList<>(configTags);
+		tags.addAll(getTagsFromMetricGroup(group));
+
+		synchronized (this) {
```
No need to synchronize since you use a `ConcurrentHashMap`.

bowenli86 replied:
sure

zentol commented on flink-metrics/flink-metrics-datadog/src/main/java/org/apache/flink/metrics/datadog/DatadogHttpReporter.java:
```diff
@@ -0,0 +1,214 @@
+				counters.put(c, new DCounter(c, name, tags));
+			} else if (metric instanceof Gauge) {
+				Gauge g = (Gauge) metric;
+				gauges.put(g, new DGauge(g, name, tags));
```
You will run into `ClassCastExceptions` here. You have to check whether the Gauge returns a number before creating the `DGauge`.

bowenli86 replied:
I made DGauge a generic one, rather than tied to any specific type. The reason being that I found sometimes the metric in this line doesn't have any value, and will throw NullPointException when calling Gauge#getValue().

The part that checks if a Gauge returns a number is in DatadogHttpReporter#report()

zentol replied:
In this case you should remove the gauge from the maps.

bowenli86 replied:
To clarify a bit more, what I found is that some Gauges might not have values ready when #notifyOfAddedMetric is called (maybe due to the value it returns has not been initialized?), but soon after will have their values available. 

Thus, I think it's good to leave this code as is

zentol replied:
I know that some metrics may return null at the beginning (although they shouldn't...). What i meant was that to remove a gauge from the map of stored gauges when you detect within report() that the gauge does not return a number.

zentol commented on flink-metrics/flink-metrics-datadog/src/main/java/org/apache/flink/metrics/datadog/DatadogHttpReporter.java:
```diff
@@ -0,0 +1,214 @@
+			} else if (metric instanceof Meter) {
+				meters.remove(metric);
+			} else if (metric instanceof Histogram) {
+				// No Histogram and Meter metrics are registered
```
comment is outdated

bowenli86 replied:
updated

zentol commented on flink-metrics/flink-metrics-datadog/src/main/java/org/apache/flink/metrics/datadog/DMeter.java:
```diff
@@ -0,0 +1,42 @@
+ * Only consider rate of the meter, due to Datadog HTTP API's limited support of meter
+ * */
+public class DMeter extends DMetric {
+	private final Meter meter;
```
The indendation in this class appears to be different than in others.

bowenli86 replied:
I double checked. They are the same using tabs. Maybe a github UI issue if you saw something different

zentol commented on flink-metrics/flink-metrics-datadog/src/main/java/org/apache/flink/metrics/datadog/DatadogHttpClient.java:
```diff
@@ -0,0 +1,83 @@
+
+	private final String seriesUrl;
+	private final String validateUrl;
+	private final OkHttpClient client;
```
you may want to add a close() method that shuts down the OkHttpClient as described in https://github.com/square/okhttp/blob/master/okhttp/src/main/java/okhttp3/OkHttpClient.java

bowenli86 replied:
You might be looking at shutting down client cache. As the doc there mentioned, "<h3>Shutdown isn't necessary</h3>" :)

zentol replied:
The documentation states "The threads and connections that are held will be released automatically if they remain idle.
But if you are writing a application that needs to aggressively release unused resources you may do so."

Since there is no harm in releasing the resources earlier, let's do that.

bowenli86 replied:
done

bowenli86 replied:
don't really like how github handles code review. This thread is hidden, and I just found it

zentol commented on flink-metrics/flink-metrics-datadog/src/test/java/org/apache/flink/metrics/datadog/DatadogHttpReporterTests.java:
```diff
@@ -0,0 +1,40 @@
+	@Test
+	public void testFilterChars() {
+		assertEquals("", reporter.filterCharacters(""));
+        assertEquals("abc", reporter.filterCharacters("abc"));
```
missing indentation

bowenli86 replied:
hmm... it's showing fine on my machine. Reindent it any way...

zentol commented on flink-metrics/flink-metrics-datadog/src/test/java/org/apache/flink/metrics/datadog/DatadogHttpReporterTests.java:
```diff
@@ -0,0 +1,40 @@
+public class DatadogHttpReporterTests {
+	private static DatadogHttpReporter reporter;
+
+	@BeforeClass
```
if there's only a single test we don't need a @BeforeClass method nor a static field; just introduce a local variable.

bowenli86 replied:
sure

zentol commented on flink-metrics/flink-metrics-datadog/src/main/java/org/apache/flink/metrics/datadog/DatadogHttpReporter.java:
```diff
@@ -0,0 +1,214 @@
+				}
+			}
+
+			for(DCounter c : counters.values()) {
```
there are several occurrences of missing spaces after `for` and `if`.

bowenli86 replied:
I'll fix them. 

But, is it enforced programmatically? I didn't see any checkstyle errors

zentol replied:
we are in the process of adding a checkstyle currently; until then we check manually for things.

zentol commented on docs/monitoring/metrics.md:
```diff
@@ -436,6 +436,35 @@ metrics.reporter.stsd.port: 8125
+metrics.reporter.dghttp.tags: myflinkapp,prod
+
+// <host>, <job_name>, <tm_id>, <subtask_index>, <task_name>, <operator_name> will be sent to Datadog as tags
+metrics.scope.jm: <host>.jobmanager
```
this section seems unnecessary

bowenli86 replied:
removed

bowenli86 replied:
simplified it and moved out of code highlight.

I believe it's necessary to explain to users how to find and filter tags in Datadog, in order to ramp them up faster and remove any potential friction.

zentol commented on flink-metrics/flink-metrics-datadog/pom.xml:
```diff
@@ -0,0 +1,79 @@
+    <groupId>org.apache.flink</groupId>
+    <artifactId>flink-metrics-datadog</artifactId>
+    <version>1.3-SNAPSHOT</version>
+
```
you should create a fat-jar using the maven-shade-plugin. If the okhttp and okio dependencies have significant dependencies as well these dependencies should eb shaded as well.

bowenli86 replied:
okhttp and okio are self contained, and they don't have any other dependencies.

can you please elaborate more "you should create a fat-jar using the maven-shade-plugin"? I have never done mvn jar shading before...

zentol replied:
Every reporter jar the is put under /opt must be self-contained, i.e. contain all dependencies that it requires. This can be done with the maven-jar-plugin, but usually prefer doing it with the maven-shade-plugin, as seen here https://github.com/zentol/flink/blob/ecda00964b46ed1203f58cb8daef9989b0892fa4/flink-metrics/flink-metrics-statsd/pom.xml.

bowenli86 replied:
the one you gave doesn't seem to work. I borrowed the one with maven-assembly-plugin in https://github.com/apache/flink/blob/master/flink-metrics/flink-metrics-graphite/pom.xml

zentol commented on flink-metrics/flink-metrics-datadog/pom.xml:
```diff
@@ -0,0 +1,79 @@
+
+		<dependency>
+			<groupId>com.fasterxml.jackson.core</groupId>
+			<artifactId>jackson-databind</artifactId>
```
You should be able to set this to provided.

bowenli86 replied:
done

zentol commented on flink-metrics/flink-metrics-datadog/src/main/java/org/apache/flink/metrics/datadog/DMetric.java:
```diff
@@ -0,0 +1,65 @@
+	}
+
+	public MetricType getType() {
+		return type;
```
what is the purpose of this field from DataDog's perspective? I don't see it being mentioned under http://docs.datadoghq.com/api/#metrics-post.

bowenli86 replied:
That's the python api, which already encapsulate metric types. I'm using the HTTP API, and it's here http://docs.datadoghq.com/api/?lang=console#metrics-post

bowenli86 replied:
In DGauge and DCounter, their fields 'gauge' and 'counter' don't impact the json serialization. 

I commented on DCounter#getMetricValue() and DGauge#getMetricValue()

zentol commented on flink-metrics/flink-metrics-datadog/src/main/java/org/apache/flink/metrics/datadog/DatadogHttpClient.java:
```diff
@@ -0,0 +1,83 @@
+
+		try {
+			Response response = client.newCall(r).execute();
+			if(!response.isSuccessful()) {
```
The response can fail for reasons other than the key being invalid, no? If so, the exception message should state "Failed to validate API key. <error code>".

bowenli86 replied:
Make sense. To broaden failure scenarios a bit

bowenli86 replied:
I can't think of any case it fails and not because invalid key. 

If network or Datadog endpoint is down, the response will just timeout, and the code path goes to catch clause.

zentol commented on flink-metrics/flink-metrics-datadog/src/main/java/org/apache/flink/metrics/datadog/DatadogHttpReporter.java:
```diff
@@ -0,0 +1,214 @@
+				request.addMeter(m);
+			}
+
+			request.send();
```
I would add a `send(DatadogHttpRequest request)` method to the DatadogHttpClient instead. Then we don't have the odd access pattern of `DatadogHttpReporter ->DatadogHttpRequest.send() -> DatadogHttpReporter.client`

bowenli86 replied:
done

zentol commented on flink-metrics/flink-metrics-datadog/src/main/java/org/apache/flink/metrics/datadog/utils/TimestampUtils.java:
```diff
@@ -0,0 +1,27 @@
+
+package org.apache.flink.metrics.datadog.utils;
+
+public class TimestampUtils {
```
This class seems overkill; could also be a utility method in the DatadogHttpClient.

bowenli86 replied:
moved to DMetric

zentol commented on flink-metrics/flink-metrics-datadog/src/main/java/org/apache/flink/metrics/datadog/utils/SerializationUtils.java:
```diff
@@ -0,0 +1,30 @@
+import com.fasterxml.jackson.core.JsonProcessingException;
+import com.fasterxml.jackson.databind.ObjectMapper;
+
+public class SerializationUtils {
```
This class seems overkill; could also be a utility method in the DatadogHttpClient.

bowenli86 replied:
moved to DatadogHttpClient

bowenli86 commented:
@zentol ready for another round. The build fails because there are a few unrelated flake tests timing out in travis build

bowenli86 commented on docs/monitoring/metrics.md:
```diff
@@ -436,6 +436,29 @@ metrics.reporter.stsd.port: 8125
+In order to use this reporter you must copy `/opt/flink-metrics-datadog-{{site.version}}.jar` into the `/lib` folder
+of your Flink distribution.
+
+Note any variables in Flink metrics, such as `<host>`, `<job_name>`, `<tm_id>`, `<subtask_index>`, `<task_name>`, and `<operator_name>`, will be sent to Datadog as tags
```
updated

zentol commented:
Could you change the PR title to "[FLINK-6013][metrics] Add Datadog HTTP metrics reporter"? The comments aren't being mirrored to JIRA.

bowenli86 commented on flink-metrics/flink-metrics-datadog/src/main/java/org/apache/flink/metrics/datadog/DatadogHttpReporter.java:
```diff
@@ -0,0 +1,200 @@
+		client = new DatadogHttpClient(config.getString(API_KEY, null));
+		LOGGER.info("Configured DatadogHttpReporter");
+
+		configTags = getTagsFromConfig(config.getString(TAGS, null));
```
Make sense. Done

bowenli86 commented on flink-metrics/flink-metrics-datadog/pom.xml:
```diff
@@ -0,0 +1,78 @@
+		<relativePath>..</relativePath>
+	</parent>
+
+    <artifactId>flink-metrics-datadog</artifactId>
```
updated

bowenli86 commented:
Where is the title that you're referring to? This PR's title at the very top is already "[FLINK-6013][metrics] Add Datadog HTTP metrics reporter" (without double quote)

zentol commented:
In the current PR title "Flink" isn't upper-case.

bowenli86 commented:
Aha! Updated! Apparently Chrome search is case insensitive

bowenli86 commented:
Green build! @zentol

zentol commented on flink-metrics/flink-metrics-datadog/src/main/java/org/apache/flink/metrics/datadog/DatadogHttpClient.java:
```diff
@@ -0,0 +1,97 @@
+
+	public DatadogHttpClient(String dgApiKey) {
+		if(dgApiKey == null || dgApiKey.isEmpty()) {
+			throw new IllegalArgumentException(
```
please remove the line-break here

bowenli86 replied:
done

zentol commented on flink-metrics/flink-metrics-datadog/src/main/java/org/apache/flink/metrics/datadog/DatadogHttpReporter.java:
```diff
@@ -0,0 +1,197 @@
+				g.getMetricValue();
+				request.addGauge(g);
+			} catch (Exception e) {
+				// ignore if the Gauge is not of Number type
```
You can safely remove the metric in this case.

bowenli86 replied:
nice catch! done

zentol commented on flink-metrics/flink-metrics-datadog/src/main/java/org/apache/flink/metrics/datadog/DatadogHttpReporter.java:
```diff
@@ -0,0 +1,197 @@
+		} else if (metric instanceof Gauge) {
+			Gauge g = (Gauge) metric;
+			gauges.put(g, new DGauge(g, name, tags));
+		} else if(metric instanceof Meter) {
```
missing space after if

bowenli86 replied:
done

bowenli86 commented on flink-dist/src/main/assemblies/opt.xml:
```diff
@@ -95,5 +95,12 @@
 		</file>
+
+		<file>
+			<source>../flink-metrics/flink-metrics-datadog/target/flink-metrics-datadog-${project.version}.jar</source>
```
Can you elaborate more, and give me more guidance on how to do it correctly? 

I copied this piece of code, and saw it exists in flink-metrics-dropwizard/pom.xml, flink-metrics-ganglia/pom.xml, and flink-metrics-graphite/pom.xml.

bowenli86 replied:
@zentol any details?

zentol replied:
The `maven-jar-plugin` creates 2 jars. For the datadog reporter these are, right now, 

1. `flink-metrics-datadog-1.3-SNAPSHOT.jar`
2. `flink-metrics-datadog-1.3-SNAPSHOT-jav-with-dependencies.jar`

The first jar doesn't contain any dependencies, whereas the second one contains all dependencies declared in the pom.xml, as long as the aren't set to `provided`. We want the latter to be placed in `/opt`, as such the `<source>` tag must point to that jar.

Basically replace the existing `<source>` configuration with `<source>../flink-metrics/flink-metrics-datadog/target/flink-metrics-datadog-${project.version}-jar-with-dependencies.jar</source>`

bowenli86 replied:
Thank you for clarifying!

bowenli86 commented on flink-metrics/flink-metrics-datadog/src/main/java/org/apache/flink/metrics/datadog/DatadogHttpClient.java:
```diff
@@ -0,0 +1,97 @@
+	private final String apiKey;
+
+	public DatadogHttpClient(String dgApiKey) {
+		if(dgApiKey == null || dgApiKey.isEmpty()) {
```
done

bowenli86 commented:
@zentol any further details and suggestions on the maven-jar-plugin part?

StephanEwen commented:
I think this is starting to look very good!

Given that this introduces new libraries as dependencies (okhttp, okio), should we pro-actively shade those away to avoid dependency conflicts?
Admittedly, it would only impact users that explicitly drop in the datadog reporter, but it might still be nice for those users. Given that we build a jr-with-dependencies anyways, the step to shading is small...

bowenli86 commented:
@StephanEwen @zentol I shaded okhttp3 and okio from flink-metrics-datadog. I didn't use '<if>shade-flink</id>' because I found it somehow prevents me from building a uber jar. Let me know if it's ok to shade them this way.


```
$ jar -tf target/flink-metrics-datadog-1.3-SNAPSHOT-jar-with-dependencies.jar

META-INF/
META-INF/MANIFEST.MF
META-INF/DEPENDENCIES
META-INF/LICENSE
META-INF/NOTICE
org/
org/apache/
org/apache/flink/
org/apache/flink/metrics/
org/apache/flink/metrics/datadog/
org/apache/flink/metrics/datadog/DatadogHttpClient.class
org/apache/flink/metrics/datadog/DatadogHttpReporter$DatadogHttpRequest.class
org/apache/flink/metrics/datadog/DatadogHttpReporter.class
org/apache/flink/metrics/datadog/DCounter.class
org/apache/flink/metrics/datadog/DGauge.class
org/apache/flink/metrics/datadog/DMeter.class
org/apache/flink/metrics/datadog/DMetric.class
org/apache/flink/metrics/datadog/DSeries.class
org/apache/flink/metrics/datadog/MetricType.class
META-INF/maven/
META-INF/maven/org.apache.flink/
META-INF/maven/org.apache.flink/flink-metrics-datadog/
META-INF/maven/org.apache.flink/flink-metrics-datadog/pom.xml
META-INF/maven/org.apache.flink/flink-metrics-datadog/pom.properties
META-INF/maven/org.apache.flink/force-shading/
META-INF/maven/org.apache.flink/force-shading/pom.xml
META-INF/maven/org.apache.flink/force-shading/pom.properties
org/apache/flink/shaded/
org/apache/flink/shaded/okhttp3/
org/apache/flink/shaded/okhttp3/Address.class
.....
org/apache/flink/shaded/okhttp3/WebSocketListener.class
META-INF/maven/com.squareup.okhttp3/
META-INF/maven/com.squareup.okhttp3/okhttp/
META-INF/maven/com.squareup.okhttp3/okhttp/pom.xml
META-INF/maven/com.squareup.okhttp3/okhttp/pom.properties
org/apache/flink/shaded/okio/
org/apache/flink/shaded/okio/AsyncTimeout$1.class
...
org/apache/flink/shaded/okio/Util.class
META-INF/maven/com.squareup.okio/
META-INF/maven/com.squareup.okio/okio/
META-INF/maven/com.squareup.okio/okio/pom.xml
META-INF/maven/com.squareup.okio/okio/pom.properties
```

StephanEwen commented:
+1 from my side
@zentol any further comments/concerns?

zentol commented on flink-metrics/flink-metrics-datadog/pom.xml:
```diff
@@ -0,0 +1,109 @@
+							<goal>shade</goal>
+						</goals>
+						<configuration>
+							<shadedClassifierName>jar-with-dependencies</shadedClassifierName>
```
I would remove this and adjust the `opt.xml` again; this makes it more consistent with other shaded jars that we produce.

bowenli86 replied:
After removing this line, the shaded jar will be named 'flink-metrics-datadog-1.3-SNAPSHOT-shaded.jar'. Is it what you want?

zentol replied:
yes.

zentol commented on flink-metrics/flink-metrics-datadog/src/main/java/org/apache/flink/metrics/datadog/DatadogHttpReporter.java:
```diff
@@ -0,0 +1,199 @@
+	private DatadogHttpClient client;
+	private List<String> configTags;
+
+	public static final String API_KEY = "apikey";
```
You could define these as a `ConfigOption` which would make it more obvious what the default value is.

bowenli86 replied:
I wouldn't worry too much about since 1) the current code works well, and changing it doesn't provide extra readability 2) it requires dependency on flink-core, which is completely unnecessary just because of a config

bowenli86 commented:
@StephanEwen @zentol 
Hi guys, I just tested it on my machine and it works as expected. What's your plan on this PR? I found on email thread that there'll be a feature freeze of release 1.3 on May 1st, and I'd really desire to get this into 1.3 :)

zentol commented:
Sorry, I'm rather busy at the moment myself due to the feature freeze and may not be able to try this out in time.

To be honest, given that if we merge it for 1.4-SNAPSHOT the jar can be easily ported to 1.3 (just replace the versions in the pom.xml and rebuild it) I'm prioritizing this less than other changes that don't have this luxury.

StephanEwen commented:
@bowenli86 Can you build and test this? Then I would be +1 to merge it to 1.3
It would not be committer tested, but contributor tested, and it is an optional component that is not part of Flink's core functionality, so less critical.

bowenli86 commented:
@StephanEwen @zentol  I've made the following changes:

1) Datadog (DD) itself has a bug of being unstable when users filter metrics by 'host' in dashboards if 'host' is sent as tags, details in [FLINK-6013](https://issues.apache.org/jira/browse/FLINK-6013). Communicated with DD engineers, they fixed the issue, and now, as advised by them,  I'm making 'host' a separate field in metric serialization. Tested the new approach successfully, and we are able to filter metrics by 'host' consistently correctly on DD dashboard.
2) Added more unit tests for the above change. Unit tests cover both cases when 'host' is present in MetricGroup and when it isn't
3) Found Maven can only recognize and run unit tests when file name ends with 'Test' rather than 'Tests'. So I renamed my test file to 'DatadogHttpClientTest' from 'DatadogHttpClientTests'
4) Upgraded okhttp and okio to newer versions
5) Found mocking system millisec will impact the new okhttp. So I separated unit tests to two enclosed test sets in 'DatadogHttpClientTest' 

Having successfully tested this across my company in the past couple days, I'm now pretty confident this is production ready. 

I agree on @zentol 's proposal if you guys are currently busy with other things, as long as we make sure this ends up in 1.3 :)  Thanks, guys!

zentol commented on flink-metrics/flink-metrics-datadog/src/test/java/org/apache/flink/metrics/datadog/DatadogHttpClientTest.java:
```diff
@@ -0,0 +1,194 @@
+	public static class TestApiKey {
+		@Test(expected = IllegalArgumentException.class)
+		public void testValidateApiKey() {
+			new DatadogHttpClient("fake_key");
```
This always sends a request to the Datadog servers correct? I'm a bit skeptical since this means if the servers are down we end up with a failing test. In general though, throwing out requests at them even though we know they will fail anyway doesn't seem ehhh.... nice?

Just checking that it fails when supplying null or an empty string is enough.

bowenli86 replied:
Make sense. There're lots of uncertainty why testing integration with 3rd party - you don't want to depend to much on it, but you have to test it somehow. It's hard to find a balance somewhere in between.

I'll make changes as you recommended.

bowenli86 commented:
@zentol do you have any other concerns?

zentol commented:
No, I think we addressed all issues :)

bowenli86 commented:
@zentol cool! Let me know how to get this into 1.3 :)

zentol commented:
This won't make it into 1.3. For one the feature freeze is in effect since today, so it is too late to merge it now.

That said it may be possible to merge it for 1.4 and maintain a separate repository for a 1.3 version that we can also link in the docs. But I'm not entirely sure.

I tried it out this morning, and it worked very well. There was an odd thing where the host field in the metric identifier wasn't shown for the taskmanager metrics but for the jobmanager, but that may hidden by Datadog automatically when the hosts match. (i.e. the first part of the metric identifier)

bowenli86 commented:
I see - by looking at the code, I saw Flink has switched to version 1.4-SNAPSHOT 12h ago. Getting it into 1.4 is fine, our company has a open ticket for getting this code into Flink, and I want to get it done. 

We haven't seen the issue you described on our side yet. I'll keep an eye on it if it comes up.

What's next for this PR? Anyone from Data Artisan will take a look at this PR, and merge it in?

zentol commented:
I can merge it sometime this week, once i found a few more things to merge alongside it.

StephanEwen commented:
@bowenli86 Th good thing is that should be able to use the reporter version 1.4-SNAPSHOT also with a Flink 1.3 release, because the reporter API is quite stable by now.

bowenli86 commented:
Sure.

BTW, that reminds me of one thing - I need to update the project version in my code from 1.3-SNAPSHOT to 1.4-SNAPSHOT.

zentol commented:
merging.

greghogan commented:
@zentol @StephanEwen why not also merge this into 1.3? There are literally no modified or removed lines of code.

zentol commented:
@greghogan My simple answer is that its feature freeze. There has to be some point in time where new features, regardless of size, stability etc. can't make it into a release. My understanding is that we decided to make the feature freeze date this point in time. When something like this is put into place it naturally follows that some contributions that are ready around the time of the FF will not be merged. Given that the FF was already delayed by a week, implicitly giving every contribution a 1 week grace period, I don't see a reason to make an exception here. It wasn't merged before the FF; it wasn't merged before the delayed FF, so it's not going into the release.

That's my take on the whole thing.

PS: Me closing this PR was not meant to imply "and that's the end of it". I just amended the commit before this discussion started and didn't remove it.

greghogan commented:
@zentol, perhaps "feature freeze" is better named "major feature freeze" and next week's "code freeze" as "minor feature freeze". My understanding is that the first cutoff is to prevent half-merged features (with "major" described as requiring multiple PRs), though master has always been quasi-releasable.

zentol commented:
According to https://cwiki.apache.org/confluence/display/FLINK/Time-based+releases the release of an RC happens when code-freeze is in effect.

greghogan commented:
@zentol yes, and the release release 16 days from now.

zentol commented:
Well we already messed with the timings of the process by extending the feature freeze for a week, so I'm going with the next best thing which is the order of events. (Feature freeze -> Code freeze -> RC)

But let's cut this short and ask the source of truth, aka the release manager.

@rmetzger Is the code-freeze in effect? If not, when will it start?

zentol commented:
Merging to 1.3 now.

bowenli86 commented:
Cool! 1.3 here we go!
